Add optional on-disk zarr caching to sensor angle generation

djhoese opened the pull request:
This adds optional on-disk zarr caching to sensor azimuth and sensor zenith and the longitude and latitude arrays needed for generating the sensor and solar angles. To do this I've added a special decorator to the `_angles.py` module and a few new configuration parameters. These new values include `cache_lonlats` and `cache_sensor_angles` and it uses the existing `cache_dir` parameter.

## Special Notes

1. As discussed on slack, the sensor angle generation (pyorbital's `get_observer_look`) depends on time. It also depends on satellite position which apparently changes frequently for some satellites. After running some tests I noticed that time has almost no effect (1e-10 maximum) whether it is changed by multiple years, days, or hours). To get around this caching will force the datetime to a single value (arbitrary). For satellite position it definitely makes a difference, I've had the caching modify the input position parameters (lon, lat, altitude) by rounding them to the nearest 10th of a degree (meters for altitude). This should prevent a large amount of extra cached arrays for something simple like floating point precision.
2. The above modification to the input arguments only happens if caching is turned on.
3. I added a `cache_clear` method on the decorator so that it is semi-easy to tell users how to clear the cache if needed in the future.
4. I found a "bug" in dask that I've filed here (https://github.com/dask/dask/issues/8380) and worked around in this PR.

## Performance comparison

Generating sensor angles either without the caching it takes about 9 seconds on my machine. If you then try to cache both the lon/lat arrays and the sensor angles (so overall redundant if you are only worried about sensor angles) then it takes ~13.4 seconds. If you then run it after it is cached it takes 0.5 seconds. All these times include calculating the nanmin and nanmax of the resulting angles. Also note for lon/lats alone it goes from ~3.2s to ~0.5s.

I thought there was an issue for this, but I can't find it. Otherwise, I just need to add documentation for this.

 - [ ] Closes #xxxx <!-- remove if there is no corresponding issue, which should only be the case for minor changes -->
 - [x] Tests added <!-- for all bug fixes or enhancements -->
 - [ ] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->

djhoese commented:
One thing to think about is that we could have sub-categories in the satpy.config structure. Something like `satpy.config.set({"optimization.cache_lonlats": True})` or `satpy.config.set({"cache.lonlats": True, "cache.sensor_angles": True})`.

mraspaud commented on doc/source/config.rst:
```diff
@@ -93,6 +93,56 @@ defaults to a different path depending on your operating system following
+generation depends on the observation time of the data and the position of the
+satellite (longitude, latitude, altitude). The angles are estimating by using
+a constant observation time for all cases (maximum ~1e-10 error) and by rounding
+satellite position coordinates to the nearest tenth degree for longitude and
```
What is the max angle error caused by this rounding?

djhoese replied:
Based on my tests that I tried to summarize on a slack last week, it seems a difference in the longitude position of -75.28 to -75.2 is a 0.094 degree difference on the sensor zenith. Now that is the wrong direction for rounding as my implemented code should have made -75.28 go to -75.3. I'll see if I can come up with a better answer for the realistic maximum longitude difference of 0.5.

pnuu replied:
```suggestion
satellite position coordinates to the nearest tenth of a degree for longitude and
```

mraspaud commented on satpy/modifiers/angles.py:
```diff
@@ -0,0 +1,266 @@
+    information is not stored.
+
+    """
+    # TODO: Consider cache management (LRU)
```
I think the docstring should mention what the behaviour is now (no limit caching, right?)

pnuu replied:
Agreed. And also mention something how it is determined if the existing cache is usable or not as there's no metadata included to check the cache validity.

mraspaud commented on satpy/modifiers/angles.py:
```diff
@@ -0,0 +1,266 @@
+
+
+def get_angles(data_arr: xr.DataArray) -> tuple[xr.DataArray, xr.DataArray, xr.DataArray, xr.DataArray]:
+    """Get sun and satellite angles to use in crefl calculations.
```
Is it still limited to crefl calculations?

mraspaud commented on satpy/modifiers/angles.py:
```diff
@@ -0,0 +1,266 @@
+    """
+    sun_angles = _get_sun_angles(data_arr)
+    sat_angles = _get_sensor_angles(data_arr)
+    # sata, satz, suna, sunz
```
Can we make explicit the variables instead and skip this comment?

mraspaud commented on satpy/modifiers/angles.py:
```diff
@@ -0,0 +1,266 @@
+    return sata, satz
+
+
+@cache_to_zarr("cache_sensor_angles", sanitize_args_func=_sanitize_observer_look_args)
```
I'd rather have two decorators here, one for argument rounding, and one for the caching if possible.

pnuu replied:
Agreed.

djhoese replied:
I don't think I can do that. We only want to round the arguments if we are caching. We can't know if we are caching unless we're in the caching logic and the caching logic can't know if it can cache unless it has the sanitized arguments (dealing with datetimes or DataArray inputs) and knows the `cache_dir` to use.

mraspaud replied:
the way I thought about is it would look like:
```
@sanitize_oberver_look_args_if("cache_sensor_angles")
@cache_to_zarr_if("cache_sensor_angles")
...
```

djhoese replied:
I like the idea, but I think it may end up being a lot of duplicated code. To purely sanitize the arguments based on this config value would be pretty easy. There is the additional complexity of `cache_dir` needing to be defined, but I guess if I assume that `cache_dir` (from `satpy.config`) is always defined then that removes that chunk of code. The only other complication then is the "uncacheable" types. The sanitizing decorator should only return sanitized arguments if the config parameter is `True` and none of the arguments are of the "uncacheable" types. This means that both the cache_to_zarr_if decorator and the sanitize decorator would have to know these types. If you're OK with these workarounds then I'm ok implementing it @mraspaud. Let me know.

Other thing to keep in mind is that this would mean that the order of the decorators matters a lot.

mraspaud replied:
If it adds a lot of complexity I don't think it's worth it. So you can leave it be.

djhoese replied:
I think it does add complexity, duplicate code, and misdirection (I'm biased as the person who wrote the original method). So I think I'm going to leave it. I take this as an approval from both of you then so this can be merged. I'll finish up another project and then merge this later in case you have other concerns.

mraspaud commented on satpy/modifiers/angles.py:
```diff
@@ -0,0 +1,266 @@
+    return satz
+
+
+@cache_to_zarr("cache_lonlats")
```
Suggestion for the name of the decorator: `cache_to_zarr_if`. This way, the signification of the argument is made clear.

pnuu replied:
To me the additional `_if` would indicate a conditional, and the supplied argument isn't a conditional but the key what to cache.

djhoese replied:
@pnuu to clarify, the key provided is the config parameter name that is the boolean of whether or not to cache. The key used in caching is based on the name of the function being decorated. So with that said, what do you think about `_if`?

mraspaud reviewed (changes requested):
Great work, I especially like the adherence to exisisting caching apis ! A few things might need to be reviewed.

pnuu commented on doc/source/config.rst:
```diff
@@ -93,6 +93,56 @@ defaults to a different path depending on your operating system following
 
+.. _config_cache_lonlats_setting:
+
+Cache Lon/Lats
+^^^^^^^^^^^^^^
```
```suggestion
Cache Longitudes and Latitudes
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

pnuu commented on doc/source/config.rst:
```diff
@@ -93,6 +93,56 @@ defaults to a different path depending on your operating system following
+* **YAML/Config Key**: ``cache_lonlats``
+* **Default**: ``False``
+
+Whether or not generating longitude and latitude coordinates should be cached
```
```suggestion
Whether or not generated longitude and latitude coordinates should be cached
```

pnuu commented on doc/source/config.rst:
```diff
@@ -93,6 +93,56 @@ defaults to a different path depending on your operating system following
+* **YAML/Config Key**: ``cache_sensor_angles``
+* **Default**: ``False``
+
+Whether or not generating sensor azimuth and sensor zenith angles should be
```
```suggestion
Whether or not generated sensor azimuth and sensor zenith angles should be
```

pnuu commented on doc/source/config.rst:
```diff
@@ -93,6 +93,56 @@ defaults to a different path depending on your operating system following
+This caching requires producing an estimate of the angles to avoid needing to
+generate new angles for every new data case. This happens because the angle
+generation depends on the observation time of the data and the position of the
+satellite (longitude, latitude, altitude). The angles are estimating by using
```
```suggestion
satellite (longitude, latitude, altitude). The angles are estimated by using
```

djhoese replied:
You know it was a rough day when the non-native english speaker is making this many corrections on the native speakers grammar.

pnuu replied:
You're wellcome :wink: :grin:

pnuu commented on satpy/modifiers/angles.py:
```diff
@@ -0,0 +1,266 @@
+# pyorbital's get_observer_look function.
+# The difference is on the order of 1e-10 at most as time changes so we force
+# it to a single time for easier caching. It is *only* used if caching.
+STATIC_EARTH_INERTIAL_DATETIME = datetime(2020, 1, 1, 0, 0, 0)
```
```suggestion
STATIC_EARTH_INERTIAL_DATETIME = datetime(2000, 1, 1, 11, 58, 55, 816000)
```
Not that it really matters, but maybe use the astronomical standard epoch J2000? See https://en.wikipedia.org/wiki/Epoch_(astronomy)#Julian_dates_and_J2000

pnuu commented on satpy/modifiers/angles.py:
```diff
@@ -0,0 +1,266 @@
+        self._uncacheable_arg_types = uncacheable_arg_types
+        self._sanitize_args_func = sanitize_args_func
+
+    def cache_clear(self, cache_dir: Optional[str] = None):
```
`cache_clear` or `clear_cache`? The latter reads better to me.

djhoese replied:
`cache_clear` matches what the `lru_cache` in functools uses: https://docs.python.org/3/library/functools.html#functools.lru_cache

pnuu reviewed (approved):
The speed-up is rather impressive!

Basically I have only few grammatical corrections and comments, so LGTM.

djhoese commented:
Ok I think I've addressed all your comments @mraspaud and @pnuu except for the one about having two decorators which I don't think is possible. Additionally while working on the docstrings I realized some things and made two additional changes:

1. I moved the "_CACHE_VERSION" inside the cache helper class so each function has its own internal version and can therefore be invalidated separately from all other decorated functions.
2. I wrapped the solar angle generation calls in a map_blocks to avoid too many dask tasks.

pnuu reviewed (approved):
LGTM

mraspaud reviewed (approved):
LGTM
